index.js

```diff
@@ -0,0 +1,130 @@
+      initialParameters.type = searchParams.get('type')
+    }
+    if (searchParams.has('adl')) {
+      // TODO Should other parameters be passed to the ADL function?
+      shouldProcessRoot = true
+      initialParameters.adl = searchParams.get('adl')
+    }
+    if (shouldProcessRoot) {
+      data = await this.#applyParameters(data, initialParameters)
+    }
+
+    // TODO: Account for dag-pb?
+    for (const { name, parameters } of segments) {
+      // This does enables ADLs to return promises for properties
+      data = await data[name]
+      const asCID = CID.asCID(data)
+      if (asCID) {
+        data = await this.getNode(asCID)
+      }
+      data = await this.#applyParameters(data, parameters)
+    }
+
+    return data
+  }
+
+  async #applyParameters (origin, { schema, adl, ...parameters }) {
+    let data = origin
+
+    const asCID = CID.asCID(data)
+    if (asCID) {
+      data = await this.getNode(asCID)
+    }
+
+    if (schema) {
+      data = await SchemaADL(data, { schema, ...parameters }, this)
+    }
+
+    if (adl) {
+      if (!this.adls.has(adl)) {
+        const known = [...this.adls.keys()].join(', ')
+        throw new Error(`Unknown ADL type ${adl}. Must be one of ${known}`)
+      }
+      data = await this.adls.get(adl)(data, parameters, this)
+    }
+
+    return data
+  }
+}
+
+export async function SchemaADL (node, { schema, type }, system) {
+  if (!schema || !type) {
+    throw new TypeError('Must specify which type to use with the schema parameter')
+  }
+
+  const schemaCID = CID.parse(schema, system.cidBases)
+  const schemaDMT = await system.getNode(schemaCID)
+  const converted = makeTyped(node, schemaDMT, type)
+
+  const typeDMT = schemaDMT.types[type]
+
+  const typedFields = new Map()
+  if (typeDMT.struct) {
+    for (const [name, fieldDMT] of Object.entries(typeDMT.struct.fields)) {
+      const nestedType = fieldDMT.type?.link?.expectedType
+      typedFields.set(name, nestedType)
+    }
+  }
+
+  if (typedFields.size) {
+    const trapped = new Proxy(converted, {
+      async get (target, property) {
+        const value = target[property]
+        if (!typedFields.has(property)) return value
+        const asCID = CID.asCID(value)
+        if (!asCID) return value
+        // Resolve the CID to the Node
+        const resolved = await system.getNode(asCID)
+        return makeTyped(resolved, schemaDMT, typedFields.get(property))
+      }
+    })
+    return trapped
+  }
+
+  // TODO: Add a trap to convert linked properties to their type, if there are subfields that are links
+
+  return converted
+}
+
+export function makeTyped (node, schemaDMT, type) {
+  const typedSchema = createTyped(schemaDMT, type)
+  const converted = typedSchema.toTyped(node)
+  if (!converted) {
+    const dataView = JSON.stringify(node)
+    const schemaDSL = toDSL(schemaDMT)
+    throw new Error(`Data did not match schema\nData: ${dataView}\nSchema:${schemaDSL}`)
+  }
```

Comment:
some fun suggestions for you to consider in addition here:

* describing the data you have as a schema so you can compare to the schema you have: https://github.com/rvagg/js-ipld-hashmap/blob/master/ipld-hashmap.js#L497-L502
* using [`@ipld/printify`](https://github.com/ipld/js-printify) to print the data instead of `JSON.stringify()` wich can get messy with bytes and CIDs
* using dag-json to print it if you just want a compact but safe form of it

Reply:
Great call on using printify!